### DISCUSSION TITLE
Apex Language Server should support methods in anon-apex #5312

### DISCUSSION THREAD

**Discussion by przemekrcloudity** [forcedotcom/salesforcedx-vscode#5834] [Ideas]
**Is your feature request related to a problem? Please describe.**

In anonymous apex scripts, defined functions aren't auto-completed. For example in code
```apex
String key(Integer x, Integer y) {
    return x + '  ' + y;
}

System.debug(key(10, 10));
```
method `key` is not autocompleted.

Whats more, parameters are not visible inside function, which means that autocompletion for both these parameters and theirs methods and properties is not working.
Variables defined inside method are visible though.

| Parameters autocompletion | Variables autocompletion |
| :-- | :-- |
| ![issue_with_anon_apex_1](https://github.com/forcedotcom/salesforcedx-vscode/assets/95613238/a7e9ee99-0e55-43ae-9359-8c87d950b9cf)  | ![issue_with_anon_apex_2](https://github.com/forcedotcom/salesforcedx-vscode/assets/95613238/51829f38-d5af-4554-b5f5-0af86039c0b4) |

**Describe the solution you'd like**

In annonymous apex scripts, I would like to have methods that were defined to be inculded in autocompletion.
In anonymous apex script, I would also like to have parameters included in autocompletion inside method.

**Describe alternatives you've considered**

Writing apex scripts inside function in apex class. THis solution is prone to error as it can be deployed by mistake

**Additional context**

Under #5312 it was suggested to convert ticket from issue to feature request

**Comment by mingxuanzhangsfdx** [comment]
Hi @przemekrcloudity thanks for creating the feature request. We will keep you and the community informed about the progress and any decisions made regarding this feature. Please feel free to reach out if you have any more ideas, questions, or feedback.

**Comment by git2gus** [comment]
This issue has been linked to a new work item: W-14837332
